subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/GradleUserHomeScopeFileTimeStampInspector.java

```diff
@@ -84,4 +89,18 @@ public void beforeComplete() {
             }
         }
     }
+
+    /**
+     * Detects whether the file system has a reliable high precision timestamp.
+     *
+     * In case both timestamp are in millisecond precision,
```

Comment:
```suggestion
     * In case any of the two timestamps is in millisecond precision,
```

Can you also describe why we can detect changes if any of the two timestamps is in millisecond precision?

I think this is because if the currentTimestamp already has millisecond precision, then all file timestamps will have millisecond precision. And the passed in `timestamp` may accidentally fall on a full second.